src/OrchardCore.Modules/OrchardCore.Localization/Startup.cs

```diff
@@ -21,7 +21,7 @@ namespace OrchardCore.Localization
     /// </summary>
     public class Startup : StartupBase
     {
-        public override int ConfigureOrder => -100;
+        public override int ConfigureOrder => 100;
```

Comment:
Changing the order makes the HttpContext.User work back. I used an arbitrary number here. Needs to be thinked more intelligently.

Reply:
Do you mean that something was no longer working before this PR? Or that you just can't insert your custom provider?

Also, just saw that a `ContentLocalization` feature also insert a provider in the options, you may need to use an higher `ConfigureOrder` in `UserLocalizationStartup` to win and not have to depend on a `ContentLocalization` feature, or use a `.PostConfigure<>()`.

Anyway will see together this night if you want, or tomorrow if I'm too tired, didn't sleep a lot last night ;)

Reply:
@Skrypt 

Okay just saw your comment

> I think I will wait for tomorrow to continue on this. It doesn't work anyways. The app.UseAuthentication(); is called after app.UseRequestLocalization(requestLocalizationOptions); which cause the HttpContext.User to always be null.

I will look at it when I will have time

Reply:
Yes, good to keep RequestLocalization early in the tenant pipeline by keeping  ConfigureOrder = -100, only Autoroute would need to be before as it is currently with the same ConfigureOrder = -100 but before because of the alphabetic order, maybe we could have used -110 but that's okay. RequestLocalization also need to be after UseRouting() but this is managed in the tenant router middleware (see below).

Looks like there is no problem to register UseAuthentication() earlier, still better to be after UseRouting() and before UseEndpoints() but this is managed in the tenant router middleware, all tenant middlewares run between UseRouting() and UseEndpoints(), the delegate passed to the last one is executed before the middleware itself is registered.

            appBuilder.UseRouting().UseEndpoints(routes =>
            {
                foreach (var startup in startups)
                {
                    startup.Configure(appBuilder, routes, ShellScope.Services);
                }
            });

So would need to be tried to see if there is no side effects, but the suggestion would be to update our `OCBuilder.AddAuthentication()` extension instead to use a negative order when we enlist UseAuthentication()

            .Configure(app =>
            {
                app.UseAuthentication();
            },
            order: -150); // Add this line

Then I saw that `UseAuthorization()` is added to the pipeline in the Users module startup without any ConfigureOrder, but I don't think that you also need this one to be run before the RequestLocalization.

Hmm, I'm also thinking about OpenId features but it seems that they don't use any middleware.

Reply:
The sign flip seems like a risky move, I'm sure it had a good reason to be `-100`. Please double-check and consider JT's most recent comment.

Also, is this only necessary for `UserLocalizationClaimsProvider`'s registration order or something else?

If the new value indeed doesn't break anything, please document it.

Reply:
Did you look into this?

Reply:
Yep, I checked it it's necessary because if we set the order for the `OC.Localization` to `-100` it will executed early on the pipeline so the user culture will not take an affect

Reply:
And did you check the other remarks above, primarily JT's? Doesn't this break anything? Also, can't we rather adjust the `Order` in the new feature's `Startup`, not to affect anything else that currently depends on Localization?

Reply:
> Also, can't we rather adjust the Order in the new feature's Startup, not to affect anything else that currently depends on Localization?

This is a very good suggestion, let me try then confirm if it works as expected - which is probably right IMHO -

Reply:
The changed code is still there but now without a comment.

Reply:
Seems I forgot to revert the changes, let me check my local changes

Reply:
Unfortunately, when `OC.User.Localization` has lower order it doesn't work too which is weird.

Reply:
Re [setting the order for `AuthenticationMiddleware`](https://github.com/OrchardCMS/OrchardCore/pull/13181/files/4d5cd8bf9187f0afa50d8faae67fb0740e25ff79..2b6a88a5572deb8e36064f3a23fcd75ed76b3f2c#diff-79fbaf725f0ce29f3982bdf05070824259b3cb3e0f5a523324151423cadef58c): 

Hmm, this is similarly impactful and might break things in an unwanted way than the `Order` of the `Startup` classes.

So can't you just change the `Order` of `OrchardCore.Users.Localization`'s `Startup`? Or perhaps its `Priority` in the manifest.

Reply:
I already tried to set it in the `OrchardCore.Users.Localization`'s `Startup` with no luck :(

Reply:
And the manifest?

Reply:
I never tried the manifest priority before, but I presume the startup order is a matter in the pipeline execution.

Let me try tomorrow, it's 2:28 AM now :)

Reply:
So, this discussion wasn't just theoretical in the end :) https://github.com/OrchardCMS/OrchardCore/pull/15173#discussion_r1551469767